src/server/tablet_snap.rs

```diff
@@ -0,0 +1,552 @@
+
+        let data = snapshot.get_data();
+        let mut snapshot_data = RaftSnapshotData::default();
+        snapshot_data.merge_from_bytes(data)?;
+        let snapshot_meta = snapshot_data.take_meta();
+        let io_type = if snapshot_meta.get_for_balance() {
+            IoType::LoadBalance
+        } else {
+            IoType::Replication
+        };
+
+        Ok(RecvTabletSnapContext {
+            key,
+            raft_msg: meta,
+            io_type,
+            start: Instant::now(),
+            chunk_size,
+        })
+    }
+
+    fn finish<R: RaftStoreRouter<impl KvEngine>>(self, raft_router: R) -> Result<()> {
+        let _with_io_type = WithIoType::new(self.io_type);
+        let key = self.key;
+        if let Err(e) = raft_router.send_raft_msg(self.raft_msg) {
+            return Err(box_err!("{} failed to send snapshot to raft: {}", key, e));
+        }
+        info!("saving all snapshot files"; "snap_key" => %key, "takes" => ?self.start.saturating_elapsed());
+        Ok(())
+    }
+}
+
+async fn send_snap_files(
+    mgr: &TabletSnapManager,
+    mut sender: impl Sink<(SnapshotChunk, WriteFlags), Error = Error> + Unpin,
+    msg: RaftMessage,
+    key: TabletSnapKey,
+    limiter: Limiter,
+) -> Result<u64> {
+    let path = mgr.get_final_path_for_gen(&key);
+    info!("begin to send snapshot file";"snap_key" => %key);
+    let files = fs::read_dir(&path)?
+        .map(|f| Ok(f?.path()))
+        .filter(|f| f.is_ok() && f.as_ref().unwrap().is_file())
+        .collect::<Result<Vec<_>>>()?;
+    let mut total_sent = msg.compute_size() as u64;
+    let mut chunk = SnapshotChunk::default();
+    chunk.set_message(msg);
+    chunk.set_data(usize::to_ne_bytes(SNAP_CHUNK_LEN).to_vec());
+    sender
+        .feed((chunk, WriteFlags::default().buffer_hint(true)))
+        .await?;
+    for path in files {
+        let name = path.file_name().unwrap().to_str().unwrap();
+        let mut buffer = Vec::with_capacity(SNAP_CHUNK_LEN);
+        buffer.push(name.len() as u8);
+        buffer.extend_from_slice(name.as_bytes());
+        let mut f = File::open(&path)?;
+        let mut off = buffer.len();
+        let mut new_len = off;
+        loop {
+            unsafe { buffer.set_len(SNAP_CHUNK_LEN) }
+            let readed = f.read(&mut buffer[off..])?;
+            // It should switch the next file if the current file read eof and the last
+            // read buffer length is less SNAP_CHUNK_LEN.
+            // Notice that, it should send empty snapshot chunk to notify the receiver
+            // switch next file if the last read buffer length is equal SNAP_CHUNK_LEN
+            if readed == 0 && new_len < SNAP_CHUNK_LEN {
+                break;
+            }
+            limiter.consume(readed);
+            new_len = readed + off;
+            total_sent += new_len as u64;
+            unsafe {
+                buffer.set_len(new_len);
+            }
+            let mut chunk = SnapshotChunk::default();
```

Comment:
This is still incorrect. When `new_len < SNAP_CHUNK_LEN` but file is not ended yet, the receiver will mistakenly treat this chunk as EOF.

Here's the full version of what I mentioned before:

```rust
for path in files {
  let mut current_chunk_len = buffer.len();
  loop {
    buffer.set_len(chunk_size);
    while current_chunk_len < chunk_size {
      let readed = f.read(&mut buffer)?;
      if readed == 0 { break; }
      total_len += readed;
    }
    /*send chunk*/
    if current_chunk_len < chunk_size {
      break;
    }
    buffer = Vec::new();
    current_chunk_len = 0;
  }
}
```